#45 5 module service

PhilippeLeopoldie opened the pull request:
🔹 ModuleRepository Updates

     -Changed GetModuleByIdAsync return type to Task<Module?> to allow null values.
     -Added GetModuleByNameAsync to IModuleRepository and implemented it in ModuleRepository.
     -Refined IModuleRepository to inherit from IRepositoryBase<Module> and IInternalRepositoryBase<Module>.

🔹 ModuleService Enhancements

    -Added IModuleService interface with full CRUD operations and registered it with lazy initialization.

    -Implemented new methods in ModuleService:
         -GetAllModulesAsync
        -GetModuleByIdAsync 
        -GetModuleByNameAsync
        -PostModuleAsync, PutModuleAsync, PatchModuleAsync, DeleteModuleAsync

    -Refactored naming conventions and improved consistency across methods.

🔹 Exception Handling

    -Added ModuleNotFoundException with overloads for ID and name.
    -Refactored NotFoundException to be abstract.
    -Introduced BadRequestException (abstract) and InvalidEntryBadRequestException.
    -Updated middleware to handle BadRequestException with proper HTTP 400 responses.

🔹 Other Improvements

    -Enhanced dependency injection setup in ServiceExtensions.

PhilippeLeopoldie committed:
Change GetModuleByIdAsync to return nullable Module

Updated the return type of `GetModuleByIdAsync` in both
`IModuleRepository` and `ModuleRepository` from `Task<Module>`
to `Task<Module?>` to allow for null returns when a module
is not found. Improved code formatting in `ModuleRepository`
for better readability.

PhilippeLeopoldie committed:
Add GetModuleByNameAsync to IModuleRepository

Updated IModuleRepository to include GetModuleByNameAsync method for retrieving a Module by name.
Implemented the method in ModuleRepository to search for a Module using the provided name and trackChanges parameter.

PhilippeLeopoldie committed:
Refactor ModuleService and add GetAllModulesAsync method

Updated the `ModuleService` class to change the private field name for the unit of work from `_unitOfWork` to `_uow`. Introduced a new asynchronous method `GetAllModulesAsync` that retrieves and maps modules to `ModuleDto` objects, returning both the DTOs and metadata.

PhilippeLeopoldie committed:
Add ModuleNotFoundException class for module errors

Introduces a new `ModuleNotFoundException` class that
inherits from `NotFoundException`. This class provides
two constructors for handling errors related to missing
modules, one for an integer `id` and another for a
string `name`, each with a descriptive error message.

PhilippeLeopoldie committed:
Add methods for retrieving modules by ID in ModuleService

- Introduced new using directives for entities and exceptions.
- Added `GetTournamentByIdAsync` to retrieve a module by ID and map it to `ModuleDto`.
- Implemented `GetTournamentByIdOrThrowExceptionAsync` for error handling when a module is not found.
- Enhances the functionality and robustness of the `ModuleService`.

PhilippeLeopoldie committed:
Refactor module-related methods for clarity

Updated method names and parameters to improve clarity and consistency.
Removed unnecessary space in `GetAllModulesAsync`.
Renamed `GetTournamentByIdAsync` to `GetModuleByIdAsync` and changed its parameter from `includeGames` to `includeActivities`.
Renamed private method `GetTournamentByIdOrThrowExceptionAsync` to `GetModuleByIdOrThrowExceptionAsync` with similar parameter updates.
Adjusted logic to ensure correct data handling and exception throwing.

PhilippeLeopoldie committed:
Add GetModuleByNameAsync method to ModuleService

Implemented a new asynchronous method `GetModuleByNameAsync` in the `ModuleService` class. This method retrieves a module by its name, checks for existence, and throws a `ModuleNotFoundException` if not found. If the module exists, it maps the module to a `ModuleDto` and returns it.

PhilippeLeopoldie committed:
Refactor NotFoundException to be abstract

PhilippeLeopoldie committed:
Add BadRequestException and InvalidEntryBadRequestException

Introduce an abstract class `BadRequestException` for handling bad request errors, featuring a `Title` property and a customizable constructor. Add `InvalidEntryBadRequestException` with two constructors for specific invalid entry scenarios.

PhilippeLeopoldie committed:
Add PutModuleAsync method to ModuleService

Implemented a new asynchronous method `PutModuleAsync` in the `ModuleService` class. This method validates the provided `id` against the `ModuleUpdateDto`, retrieves the corresponding module, updates its properties, and saves the changes to the database.

PhilippeLeopoldie committed:
Add methods for patching modules in ModuleService

Implemented `ModuleToPatchAsync` to retrieve a module and its DTO, and `SavePatchModuleAsync` to apply updates and save changes asynchronously.

PhilippeLeopoldie committed:
Enhance IModuleRepository with interface inheritance

Updated the IModuleRepository interface to extend IRepositoryBase<Module> and IInternalRepositoryBase<Module>. This change improves functionality and clarifies the contract for module-related operations while keeping existing method signatures intact.

PhilippeLeopoldie committed:
Add PostModuleAsync method to ModuleService

Implemented a new asynchronous method `PostModuleAsync` in the `ModuleService` class. This method accepts a `ModuleCreateDto`, maps it to a `Module` entity, creates the module in the repository, and returns a `ModuleDto` after completing the unit of work.

PhilippeLeopoldie committed:
Add DeleteModuleAsync method to ModuleService

Implemented a new asynchronous method `DeleteModuleAsync` in the `ModuleService` class. This method allows for the deletion of a module by its ID, retrieving the module with activities and tracking changes before deletion, and completing the unit of work asynchronously.

PhilippeLeopoldie committed:
Add IModuleService for module management functionality

This commit introduces a new service interface, `IModuleService`, and its implementation in the `ModuleService` class. The `IModuleService` is registered in `ServiceExtensions.cs`, enhancing the dependency injection setup. The interface defines methods for module management, including deletion, retrieval, and updates, improving the modularity and maintainability of the codebase.

PhilippeLeopoldie committed:
Add lazy registration for IModuleService

This commit introduces a new service registration for `IModuleService` in the `ServiceExtensions` class using the `AddLazy` method. This change allows `IModuleService` to be lazily instantiated alongside existing services `IAuthService` and `IActivityService`.

PhilippeLeopoldie committed:
Add BadRequestException handling to middleware

This commit introduces a new case in the ExceptionMiddlewareExtensions class to handle BadRequestException. When triggered, it sets the HTTP status code to 400 and creates a ProblemDetails object with relevant exception details. This improvement enhances error handling and provides clearer feedback for bad requests.

ChornaIryna commented on LMS.Infrastructure/Repositories/ModuleRepository.cs:
```diff
@@ -28,12 +28,19 @@ public async Task<PagedList<Module>> GetModulesAsync(
 
+    public async Task<Module?> GetModuleByNameAsync(string name, bool trackChanges)
+    {
+        return await FindByCondition(Tournament => string.Equals(Tournament.Name, name), trackChanges)
```
Kanske är 'modul' ett bättre val här. :)

PhilippeLeopoldie replied:
absolut!

JorgenHagglund commented on LMS.Services/ModuleService.cs:
Jag kan ju tycka att Create/Update/Patch/Delete är bättre namn på metoderna. Post och Put är HTTP-metod

PhilippeLeopoldie replied:
Gjort! tack :)

ChornaIryna commented on LMS.Services/ModuleService.cs:
Det kan vara bra att lägga in en kontroll så att det alltid finns en Course, och att den är kopplad till en Module.

PhilippeLeopoldie replied:
Eftersom Course redan är markerad som required i Module-entiteten och CourseId är en icke-nullbar foreign key, garanterar databasen att varje modul alltid är kopplad till en kurs. Därför behövs ingen extra kontroll i koden. Vad tänker du?

PhilippeLeopoldie committed:
Refactor parameter name in GetModuleByNameAsync

Correct typo

PhilippeLeopoldie committed:
Refactor method names in ModuleService and IModuleService

Renamed several methods in the `ModuleService` class for improved clarity and consistency. The methods `PutModuleAsync`, `ModuleToPatchAsync`, `SavePatchModuleAsync`, and `PostModuleAsync` have been changed to `UpdateModuleAsync`, `GetModuleForPatchAsync`, `ApplyModulePatchAsync`, and `CreateModuleAsync`.

Updated the `IModuleService` interface to reflect these changes, ensuring consistency between the interface and its implementation.

JorgenHagglund commented:
Ser bra ut, tycker jag!

PhilippeLeopoldie committed:
Merge branch 'development' into #45-5-Module-Service
